scripts/irods/upgrade_configuration.py

```diff
@@ -92,70 +102,60 @@ def upgrade(irods_config):
                                 'Please resolve this naming conflict manually and try again.' % (old_path, new_path, new_path))
                     shutil.move(old_path, new_path)
 
-    if old_version_tuple < (4, 3, 0):
-        # Load server_config.json.
-        with open(paths.server_config_path()) as f:
-            server_config = json.load(f)
+    if old_version_tuple[:2] <= (4, 3):
+        # Build a new server_config.json file using server_config.json.template as a base.
+        # Overwrite all configuration properties in the template with the properties from
+        # server_config.json.
+        server_config = load_server_config_with_explicit_defaults()
```

Comment:
How do we feel about this logic?

It attempts to always provide a server config that contains all options explicitly defined. This code path is always invoked when dealing with a 4.3+ server. Devs are expected to update `server_config` from here on.

A better solution is probably to go on and write `server_config` to disk. Then the code is free to reload the config and make changes as needed.

Reply:
Interesting - so the approach is to populate the template, rather than 'sniff and inject' into the existing file content.

I'm most concerned about missing things that we don't know about... like environment variables that someone might be depending on.

Building on what they already have in place is a safer starting point, I think...

Reply:
Are you talking about environment variables defined in server_config.json?

If so, this should work. We start with the template and overwrite every matching key in the local server config.

All options the admin set originally should be maintained.

Reply:
I've adjusted the logic so that it is easier to maintain and understand.